### PR TITLE
Block editor: pass patterns selector as setting

### PR DESCRIPTION
**Pull request by ellatrix**
## What?

Changes the way patterns are loaded in the block editor to a select function that resolves the patterns in the core store instead of resolving them in the block editor store.

## Why?

Currently, when loading previews with patterns, these previews all refetch the patterns from the server because these are all separate block editor instances.

image

## Testing Instructions

Patterns in the inserter should appear. Maybe worth testing other pattern lists in the site editor, but nothing should have changed there, everything should work as before.

We have e2e test coverage for pattern screens.

### CODE REVIEW

**Comment by jsnajdr** [comment]
There are unit test failures: tests need to run in an environment where `state.settings.__experimentalSelectBlockPatterns` is set.

**Comment by ellatrix** [comment]
Btw, I'm working on a new performance test first so we can track the improvements better

**Comment by Mamaduka** [comment]
@jsnajdr, I just re-tested #58665 on this branch, and it seems to fix the issue.

**Comment by ellatrix** [comment]
Awesome, will polish this PR tomorrow

**Comment by ellatrix** [comment]
@jsnajdr @Mamaduka Does this look good to you?

File: packages/editor/src/components/provider/use-block-editor-settings.js [hunk 1]
```diff
@@ -247,17 +251,8 @@ function useBlockEditorSettings( settings, postType, postId ) {
...
-				return ( await fetchBlockPatterns() ).filter(
-					( { postTypes } ) => {
-						return (
-							! postTypes ||
-							( Array.isArray( postTypes ) &&
-								postTypes.includes( postType ) )
-						);
-					}
-				);
-			},
```
Comment (ellatrix) [inline packages/editor/src/components/provider/use-block-editor-settings.js:260]: Oh, I need to move this to into a selector

**Comment by Mamaduka** [comment]
@ellatrix, can you update the PR with testing instructions?

**Comment by ellatrix** [comment]
@Mamaduka I added something, but this PR shouldn't change anything, so just making sure there's no e2e test failures.

**Comment by ellatrix** [comment]
Even if it the other bug _appears_ fixed (nothing has changed there that could have fixed it I think), this PR is still needed to avoid fetching all patterns from the server for every preview that uses the pattern block (or any other selector that depends on a patterns store selector).

**Review by jsnajdr — APPROVED** [review]
Looks good :+1:

**Review by Mamaduka — APPROVED** [review]
Makes sense. Thank you, @ellatrix!

**Comment by ellatrix** [comment]
Thank you both :)

**Comment by ellatrix** [comment]
Looking at the performance graph, there's a 10% improvement for the `Replace template modal (Site Editor)` metric, so worth back porting imo.

**Comment by ellatrix** [comment]
(Which was a regression btw)

**Comment by youknowriad** [comment]
I just cherry-picked this PR to the more/backports-for-beta3 branch to get it included in the next release: 4b25a8869ec

**Comment by creativecoder** [comment]
Note that as part of curating the Gutenberg 17.8 changelog, I added the label [[Feature] Patterns](https://github.com/WordPress/gutenberg/labels/%5BFeature%5D%20Patterns) to this PR to better categorize it.

Please let me know if there is a different label that would be a better fit.

**Pull request merged**